src/esp/gfx/PBRDrawable.cpp

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+#include "PBRDrawable.h"
+
+#include <Corrade/Containers/ArrayViewStl.h>
+#include <Corrade/Utility/FormatStl.h>
+
+namespace Mn = Magnum;
+
+namespace esp {
+namespace gfx {
+
+PBRDrawable::PBRDrawable(scene::SceneNode& node,
+                         Mn::GL::Mesh& mesh,
+                         ShaderManager& shaderManager,
+                         const Mn::ResourceKey& lightSetupKey,
+                         const Mn::ResourceKey& materialDataKey,
+                         DrawableGroup* group)
+    : Drawable{node, mesh, group},
+      shaderManager_{shaderManager},
+      lightSetup_{shaderManager.get<LightSetup>(lightSetupKey)},
+      materialData_{
+          shaderManager.get<MaterialData, PBRMaterialData>(materialDataKey)} {
+  updateShader().updateShaderLightParameters();
+}
+
+void PBRDrawable::setLightSetup(const Mn::ResourceKey& lightSetup) {
+  lightSetup_ = shaderManager_.get<LightSetup>(lightSetup);
+  updateShader().updateShaderLightParameters();
+}
+
+void PBRDrawable::draw(const Mn::Matrix4& transformationMatrix,
+                       Mn::SceneGraph::Camera3D& camera) {
+  // no need to call updateShaderLightParameters() in the draw loop,
+  // only need to update the position or direction of the lights
+  updateShader().updateShaderLightDirectionParameters(transformationMatrix,
+                                                      camera);
+
+  (*shader_)
+      // e.g., semantic mesh has its own per vertex annotation, which has been
+      // uploaded to GPU so simply pass 0 to the uniform "objectId" in the
+      // fragment shader
+      .setObjectId(
+          static_cast<RenderCamera&>(camera).useDrawableIds()
+              ? drawableId_
+              : (materialData_->perVertexObjectId ? 0 : node_.getSemanticId()))
+      .setTransformationMatrix(transformationMatrix)  // modelview matrix
+      .setMVPMatrix(camera.projectionMatrix() * transformationMatrix)
+      // because we actually do not have scaling, only rotation, so that the
+      // modelview matrix mv = inv(mv^T) where mv stands for the up-left 3x3
+      // part of the modelview matrix
+      .setNormalMatrix(transformationMatrix.rotationScaling())
+      .bindTextures(
+          materialData_->albedoTexture, materialData_->roughnessTexture,
+          materialData_->metallicTexture, materialData_->normalTexture)
+      .setBaseColor(materialData_->baseColor)
+      .setRoughness(materialData_->roughness)
+      .setMetallic(materialData_->metallic);
+
+  if (materialData_->textureMatrix != Mn::Matrix3{}) {
+    // TODO
+  }
```

Comment:
I'd add a temporary assert here -- because when MeshOptimizer comes to play, this will be the case always, and models that don't have their texture transform respected will look *extremely wrong* and nobody would know why :)

Same for the TODO below.

Reply:
MeshOptimizer quantizes texture coordinates to save memory. So for example, when a mesh uses a [0.5, 0.75] range of the texture, it'll map that range to 0-65535 and store that in a 16-byte integer instead of a 32-bit float, saving half of memory for that attribute. Then, in order to unpack that quantized coordinate back to [0.5, 0.75], it'll supply a texture matrix that scales and offsets the coordinates again. (Same is done for positions, but there this extra dequantizing transform is added to the node transform, so no extra feature is needed.)

A good model to test with is https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/Lantern/glTF-Quantized -- unless this got broken since #550, the current viewer should load and display it correctly. With texture transform missing the textures would be significantly off.

This is not a *mandatory* feature of meshopt, but it's on by default, so I'm just saying -- better have a temporary assertion in place so you know what's to blame if such a model gets loaded ;)